### DISCUSSION TITLE
Plugin for calendar (ICS) files

### DISCUSSION THREAD

**Discussion by Nick222** [zim-desktop-wiki/zim-desktop-wiki#2053] [Ideas]
Are there any plugin for Zim to import or view in Journal calendar (ICS) files ?

**Comment by Fjanks** [comment]
I was just searching for the same thing and found this issue. It seems that there is nothing like that yet. Could someone give a hint how it could be implemented? In particular, how to parse the document structure of the wiki and add or display a line of text in a specified section? Is there a plugin which does something similar?

**Reply by introt** [reply]
The Journal plugin has relevant code for this task (clicking on the calendar widget takes you to the relevant page regardless of whether you're using a page per day or week) - otherwise, see the notebook API: https://introt.github.io/zim-desktop-wiki/docs/api/zim.notebook.html

**Reply by introt** [reply]
For a tool working in the opposite direction, see  ZIM Task to Calendar #2654

**Comment by Nick222** [comment]
Move to Discussion, pls...

**Comment by sim-plex** [comment]
I have developed an initial version of a plugin designed for adding ICS events as tasks to the journal. This plugin currently has limited capabilities due to its early stages in development; however if there is still interest, I can further develop and publish it.

**Comment by Nick222** [comment]
Can I see initial version?

**Comment by sim-plex** [comment]
https://codeberg.org/sim-plex/icsimport

After installation, it can be used at Tools->ICS Import

**Comment by Nick222** [comment]
I attempt to import ICS with January 2025 item.

> This is zim 0.75.2
> Platform: posix
> Locale: ru_RU UTF-8
> FS encoding: utf-8
> Python: (3, 12, 7, 'final', 0)
> PyGObject: (3, 48, 2)
> 
> ======= Traceback =======
>   File "/usr/lib/python3/dist-packages/zim/actions.py", line 159, in _on_activate
>     self.__call__()
>   File "/usr/lib/python3/dist-packages/zim/actions.py", line 130, in __call__
>     return self._action.func(self._instance, *args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/nick/.local/share/zim/plugins/icsimport/icsimport.py", line 56, in execute_import
>     icsimport.add_events()
>   File "/home/nick/.local/share/zim/plugins/icsimport/icsimport.py", line 97, in add_events
>     extension_tree = self.get_extension_tree(vitems_by_date)
>                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/nick/.local/share/zim/plugins/icsimport/icsimport.py", line 114, in get_extension_tree
>     if description := vitem['DESCRIPTION']:
>                       ~~~~~^^^^^^^^^^^^^^^
>   File "/usr/lib/python3/dist-packages/icalendar/caselessdict.py", line 40, in __getitem__
>     return super().__getitem__(key.upper())
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> KeyError: 'DESCRIPTION'

**Comment by Nick222** [comment]
I see - it cannot import when this day in journal already exist

**Comment by sim-plex** [comment]
The problem was that the event needed a description. If it fixed it, please confirm. I am aware that there is more functionality missing and bugs. As I said, it's at an incomplete stage.

**Comment by Nick222** [comment]
New version worked good :)

But it import events which have already been imported - it can be prevent ?

**Reply by sim-plex** [reply]
I have created [an issue](https://codeberg.org/sim-plex/icsimport/issues/1) and would appreciate feedback.

**Reply by introt** [reply]
In addition to comparing existing "sub-trees" to new ones you could use the event UUID (if one exists) to update the sub-trees in place. You can hide the UUID inside a Zim anchor (or a link).

You could also delegate merging changes to another plugin (#2569)
